tests/unit/test_futures.py

```diff
@@ -480,24 +486,134 @@ def test_shutdown_no_wait(self):
+        def return_foo():
+            return 'foo'
+
+        worker = ExecutorWorker(self.task_queue)
+        worker.start()
+        future = ExecutorFuture()
+        self.task_queue.put(ExecutorWorkerTask(future, return_foo))
+        self.task_queue.put(ExecutorWorkerShutdownTask())
+        worker.join()
+        self.assertEqual(future.result(), 'foo')
+
+    def test_sets_exception(self):
+        def raise_runtime_error():
+            raise RuntimeError()
+
+        worker = ExecutorWorker(self.task_queue)
+        worker.start()
+        future = ExecutorFuture()
+        self.task_queue.put(ExecutorWorkerTask(future, raise_runtime_error))
+        self.task_queue.put(ExecutorWorkerShutdownTask())
+        worker.join()
+        with self.assertRaises(RuntimeError):
+            future.result()
+
+
 class TestExecutorFuture(unittest.TestCase):
+    def setUp(self):
+        self.future = ExecutorFuture()
+
+    def _get_exc_info(self, exception):
+        try:
+            raise exception
+        except type(exception):
+            return sys.exc_info()
+
+    def test_done_starts_false(self):
+        self.assertFalse(self.future.done())
+
+    def test_done_after_setting_result(self):
+        self.future.set_result('result')
+        self.assertTrue(self.future.done())
+
+    def test_done_after_setting_exception(self):
+        self.future.set_exception_info(Exception(), None)
+        self.assertTrue(self.future.done())
+
     def test_result(self):
-        with ThreadPoolExecutor(max_workers=1) as executor:
-            future = executor.submit(return_call_args, 'foo', biz='baz')
-            wrapped_future = ExecutorFuture(future)
-        self.assertEqual(wrapped_future.result(), (('foo',), {'biz': 'baz'}))
+        self.future.set_result('result')
+        self.assertEqual(self.future.result(), 'result')
 
-    def test_done(self):
-        with ThreadPoolExecutor(max_workers=1) as executor:
-            future = executor.submit(return_call_args, 'foo', biz='baz')
-            wrapped_future = ExecutorFuture(future)
-        self.assertTrue(wrapped_future.done())
+    def test_exception_result(self):
+        exception = ValueError('message')
+        self.future.set_exception_info(exception, None)
+        with self.assertRaisesRegexp(ValueError, 'message'):
+            self.future.result()
+
+    def test_exception_result_doesnt_modify_last_frame(self):
+        exception = ValueError('message')
+        tb = self._get_exc_info(exception)[2]
+        self.future.set_exception_info(exception, tb)
+        try:
+            self.future.result()
+            # An exception should have been raised
+            self.fail()
+        except ValueError:
+            actual_tb = sys.exc_info()[2]
+            last_frame = traceback.extract_tb(actual_tb)[-1]
+            last_expected_frame = traceback.extract_tb(tb)[-1]
+            self.assertEqual(last_frame, last_expected_frame)
+
+    def test_done_callback(self):
```

Comment:
I think that we should add a test to make sure an exception can be raised in the done callback and it does not affect the thread calling it (i.e. catches the exception).